### PR TITLE
MDEV-34316 Ignore the NOCOPY keyword in stored routine parameters when sql_mode=ORACLE

### PR DESCRIPTION
**Pull request by wong-github1**
- [x] *The Jira issue number for this PR is: MDEV-34316

## Description
During sql_mode=ORACLE, ignore the NOCOPY keyword in stored routineparameters. The optimization (pass-by-reference instead of pass-by-value) helping to avoid value copying will be done in a separate task when needed.

## Release Notes
Ignore the NOCOPY keyword in stored routine parameters when sql_mode=ORACLE

## How can this PR be tested?

mysql-test/mtr sp-ignore_nocopy

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

### CODE REVIEW

**Comment by LinuxJedi** [comment]
Hi @wong-github1,

Unfortunately this is causing test failures, this is because the `NOCOPY` keyword is also used for `ALTER TABLE`, for example:

```sql
ALTER TABLE t1 MODIFY d DOUBLE DEFAULT 10, ALGORITHM=NOCOPY
```

**Comment by wong-github1** [comment]
> Hi @wong-github1,
> 
> Unfortunately this is causing test failures, this is because the `NOCOPY` keyword is also used for `ALTER TABLE`, for example:
> 
> ```sql
> ALTER TABLE t1 MODIFY d DOUBLE DEFAULT 10, ALGORITHM=NOCOPY
> ```

Hi @LinuxJedi (Andres),

Thanks you so much in pointing out the error. I have just fixed it.

**Review by FooBarrior — CHANGES_REQUESTED** [review]
Hello @wong-github1! Thank you for your efforts. 

For the ORACLE compatibility mode we have an compat/oracle suite. I think you can put your tests in the sp-param.test file in that suite.

Also I think we can avoid an extra switch in your ALTER TABLE fix. Can you please update it according to my suggestion?

Kind regards.

File: sql/sql_yacc.yy [hunk 1]
```diff
@@ -1466,6 +1467,8 @@ bool my_yyoverflow(short **a, YYSTYPE **b, size_t *yystacksize);
         opt_recursive opt_format_xid opt_for_portion_of_time_clause
         ignorability
 
+%type <num> alter_algorithm_option_is_kwd
```
[the remark this answers is not in the document]
Comment (wong-github1) [reply sql/sql_yacc.yy]: Updated.

File: sql/sql_yacc.yy [hunk 2]
```diff
@@ -7939,11 +7942,28 @@ opt_index_lock_algorithm:
         | alter_algorithm_option alter_lock_option
         ;
 
+alter_algorithm_option_is_kwd:
+           DEFAULT    { $$ = 0; }
```
[the remark this answers is not in the document]
Comment (wong-github1) [reply sql/sql_yacc.yy]: Updated.

**Comment by wong-github1** [comment]
Hi @FooBarrior ,
Thanks for your review. I have made the corresponding revision.
Need your help to review again and once again, thank you.

**Review by FooBarrior — APPROVED** [review]

**Comment by FooBarrior** [comment]
Hello again @wong-github1! 

Our QA has found a bug within this patch: [MDEV-35229](https://jira.mariadb.org/browse/MDEV-35229). Can you please take a look?

The bug is serious and the priority is high, so would be nice if you can take a look as soon as possible.

If any questions, reach me out in MariaDB's zulip:
https://mariadb.zulipchat.com/#user/127697

Best regards

Nikita

**Pull request merged**